flink/src/main/java/org/apache/iceberg/flink/data/FlinkParquetReaders.java

```diff
@@ -19,64 +19,703 @@
+      } else {
+        return new ReusableMapData();
+      }
+    }
+
+    @Override
+    @SuppressWarnings("unchecked")
+    protected Map.Entry<K, V> getPair(ReusableMapData map) {
+      Map.Entry<K, V> kv = nullEntry;
+      if (readPos < map.capacity()) {
+        entry.set((K) map.keys.values[readPos], (V) map.values.values[readPos]);
+        kv = entry;
+      }
+
+      readPos += 1;
+
+      return kv;
+    }
+
+    @Override
+    protected void addPair(ReusableMapData map, K key, V value) {
+      if (writePos >= map.capacity()) {
+        map.grow();
+      }
+
+      map.keys.values[writePos] = key;
+      map.values.values[writePos] = value;
+
+      writePos += 1;
+    }
 
-    RowReader(List<Type> types, List<ParquetValueReader<?>> readers, Types.StructType struct) {
+    @Override
+    protected MapData buildMap(ReusableMapData map) {
+      map.setNumElements(writePos);
+      return map;
+    }
+  }
+
+  private static class RowDataReader extends ParquetValueReaders.StructReader<RowData, GenericRowData> {
+    private final int numFields;
+
+    RowDataReader(List<Type> types, List<ParquetValueReader<?>> readers) {
       super(types, readers);
-      this.structType = struct;
+      this.numFields = readers.size();
     }
 
     @Override
-    protected Row newStructData(Row reuse) {
-      if (reuse != null) {
-        return reuse;
+    protected GenericRowData newStructData(RowData reuse) {
+      if (reuse instanceof GenericRowData) {
+        return (GenericRowData) reuse;
       } else {
-        return new Row(structType.fields().size());
+        return new GenericRowData(numFields);
       }
     }
 
     @Override
-    protected Object getField(Row row, int pos) {
-      return row.getField(pos);
+    protected Object getField(GenericRowData intermediate, int pos) {
+      return intermediate.getField(pos);
     }
 
     @Override
-    protected Row buildStruct(Row row) {
-      return row;
+    protected RowData buildStruct(GenericRowData struct) {
+      return struct;
     }
 
     @Override
-    protected void set(Row row, int pos, Object value) {
+    protected void set(GenericRowData row, int pos, Object value) {
       row.setField(pos, value);
     }
+
+    @Override
+    protected void setNull(GenericRowData row, int pos) {
+      row.setField(pos, null);
+    }
+
+    @Override
+    protected void setBoolean(GenericRowData row, int pos, boolean value) {
+      row.setField(pos, value);
+    }
+
+    @Override
+    protected void setInteger(GenericRowData row, int pos, int value) {
+      row.setField(pos, value);
+    }
+
+    @Override
+    protected void setLong(GenericRowData row, int pos, long value) {
+      row.setField(pos, value);
+    }
+
+    @Override
+    protected void setFloat(GenericRowData row, int pos, float value) {
+      row.setField(pos, value);
+    }
+
+    @Override
+    protected void setDouble(GenericRowData row, int pos, double value) {
+      row.setField(pos, value);
+    }
```

Comment:
For flink `GenericRowData`,  we don't provide setInteger/setLong/setFloat interfaces etc, so seems we don't have to override those methods from `ParquetValueReaders.StructReader`, because all of them have the default implementation, for example: 

```java
    protected void setBoolean(I struct, int pos, boolean value) {
      set(struct, pos, value);
    }
```

I mean override the `set(I struct, int pos, Object value)` is enough.

BTW, I saw Setter interface in `ParquetValueReaders.StructReader` don't have meaningful usage. Is it necessary to create a separate pull request to remove all methods related to `Setter`  ?  cc @rdblue .